[MRG] add gated attention layer, tests, and docs

nelson-liu opened the pull request:
Added the gated attention layer described in [Dhingra et. al 2016](https://www.semanticscholar.org/paper/Gated-Attention-Readers-for-Text-Comprehension-Dhingra-Liu/200594f44c5618fa4121be7197c115f78e6e110f)

Built docs: 
API Summary for layers.attention: https://133-47793051-gh.circle-artifacts.com/0/home/ubuntu/deep_qa/doc/_build/html/layers/attention.html#gatedattention
Full API Reference for layers.attention: https://133-47793051-gh.circle-artifacts.com/0/home/ubuntu/deep_qa/doc/_build/html/api_doc/deep_qa.layers.attention.html#module-deep_qa.layers.attention.gated_attention

DeNeutoy commented on src/main/python/deep_qa/layers/attention/gated_attention.py:
```diff
@@ -0,0 +1,99 @@
+        # which is of shape (batch, document length, biGRU hidden length)
+        if self.gating_function == "*":
+            return question_update * document_matrix
+        if self.gating_function == "+":
```
Here, I think we need to utilise the mask from the document matrix. In the multiplication case, it's fine as the question update already has zeros where they should be, but for the concatenation and addition cases, if we don't mask here and don't pass a mask forward from this layer, I don't think this is going to be correctly masked upstream.

DeNeutoy commented on src/test/python/deep_qa_test/layers/attention/gated_attention_test.py:
```diff
@@ -0,0 +1,102 @@
+# pylint: disable=no-self-use,invalid-name
```
Might be good to have a test for the masked case as well, given I'm not 100% sure it would pass in the addition and concatenation cases.

matt-gardner commented on src/main/python/deep_qa/layers/attention/gated_attention.py:
```diff
@@ -0,0 +1,99 @@
+    def compute_mask(self, inputs, mask=None):
+        # pylint: disable=unused-argument
+        # We do not need a mask beyond this layer.
+        return None
```
I second what Mark said.  In addition, I think you need to return a mask here, especially seeing how you use this in the model.  After the first run of the loop, you'll no longer have a mask.

nelson-liu commented:
thanks for the reviews, @DeNeutoy and @matt-gardner . I've addressed the comments, and added tests for all three masked cases. For the `compute_mask` method, I think the right thing to do is just re-return the mask of the document; thoughts?

matt-gardner commented:
Yes, returning the document mask seems like the right thing to do here.
